state/application.go

```diff
@@ -3861,13 +3862,16 @@ func (st *State) WatchApplicationsWithPendingCharms() StringsWatcher {
 			}
 
 			// We need an application with both a charm URL and
-			// an origin set.trusty
+			// an origin set.
 			app, _ := st.Application(st.localID(sKey))
-			if app == nil || app.CharmOrigin() == nil {
+			if app == nil {
 				return false
 			}
-
-			return app.CharmPendingToBeDownloaded()
+			origin := app.CharmOrigin()
+			if origin == nil {
+				return false
+			}
+			return app.CharmPendingToBeDownloaded() || origin.Source == "charm-hub" && origin.ID == ""
```

Comment:
I wonder if `CharmPendingToBeDownloaded` should actually have the `|| origin.Source == "charm-hub" && origin.ID == ""` with in it? 

As we only call `CharmPendingToBeDownloaded` when downloading.

Reply:
It's also called by the caasapplicationprovisioner facade. I'll move the check.